ReactiveCocoa/NSObject+Lifetime.swift

```diff
@@ -20,66 +17,57 @@ extension Reactive where Base: NSObject {
 
 			// Swizzle `-dealloc` so that the lifetime token is released at the
 			// beginning of the deallocation chain, and only after the KVO `-dealloc`.
-			objc_sync_enter(objcClass)
-
-			// Swizzle the class only if it has not been swizzled before.
-			if objc_getAssociatedObject(objcClass, &lifetimeKey) == nil {
-				objc_setAssociatedObject(objcClass, &lifetimeKey, true, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
-
-				var existingImpl: IMP? = nil
-
-				let newImplBlock: @convention(block) (UnsafeRawPointer) -> Void = { objectRef in
-					// A custom trampoline of `objc_setAssociatedObject` is used, since
-					// the imported version has been inserted with ARC calls that would
-					// mess with the object deallocation chain.
-
-					// Release the lifetime token.
-					_rac_objc_setAssociatedObject(objectRef, &lifetimeTokenKey, nil, .OBJC_ASSOCIATION_RETAIN_NONATOMIC)
-
-					let impl: IMP
-
-					// Call the existing implementation if one has been caught. Otherwise,
-					// call the one first available in the superclass hierarchy.
-					if let existingImpl = existingImpl {
-						impl = existingImpl
-					} else {
-						let superclass: AnyClass = class_getSuperclass(objcClass)
-						impl = class_getMethodImplementation(superclass, deallocSelector)
+			synchronized(objcClass) {
+				// Swizzle the class only if it has not been swizzled before.
+				if objc_getAssociatedObject(objcClass, AssociationKey.lifetime) == nil {
```

Comment:
Seems like we could use custom methods on `AnyClass` too to have nicer APIs for this.